### ISSUE TITLE
Links to self are blocked

### ISSUE THREAD

**Issue by RubenFixit** [PrivateBin/PrivateBin#959]
## Steps to reproduce
1. Include a link to your PrivateBin in the bin editor or comment
2. Click on the link once sent or in preview

### What happens
Browser opens a new tab with a BLOCKED error (ERR_BLOCKED_BY_RESPONSE). 

### What should happen
Browser opens up a new tab with your PrivateBin page

## Additional information
This error occurs on both chrome and firefox.

A workaround for now is to middle click your PrivateBin links, then it works as expected.

Also, the link that is generated after posting works just fine... 

## Basic information
**Server address**: https://pastebin.tcloud.dev

**Server OS:** Ubuntu 20.04.3 LTS in a docker container ([cloudron](https://cloudron.io))

**Webserver:** nginx/1.18.0

**Browser:** All

**PrivateBin version:** 1.6.0

I can reproduce this issue on <https://privatebin.net>: Yes

Here is an example: https://privatebin.net/?8c0a4fe6d179f5a0#DmaksTnxDTYRwx3j2XjbP6uq4hWQjirTG1ZQwS5KZmUY
_**Will expire on Aug 12, 2022**_

**Comment by elrido** [comment]
Hi there, doing some experimentations this is due to the content-security-policy header that we set. You can allow this by setting the `default-src` to `'self'` instead of our recommended value of `'none'` in your `conf.php` file. I've temporarily done this on this instance, where the link to itself currently works: https://snip.dssr.ch/?72eaf59fb05b8584#7h27hKxFQngvML2ELCd1MuKc2Fg15Uv6fL1TrB4Zap42

I'd like to hear some opinions from other maintainers on this - I feel we should still leave the `default-src` at `'none'` as it is a safe fallback, and we are forced to enable access for each feature we actually do use. Server admins can opt to enable self-links via the configuration file, if they want to let people add pastes/comments with links to other pastes/discussions on the same instance.

**Comment by rugk** [comment]
Yes default-src needs to be none, but I still see that we should enable that feature. After all, some other CSP directive (that we do not explicitly specify) needs to be there then that can specify that…

I am.somewhat confused because I looked through https://content-security-policy.com/ and found no one that affects just pure links.

I know there was a header related to the same origin policy once that could break links from being opened or at least cookies from being read then(?), but I don't get the name of it now and maybe its also the wrong thing.

https://securityheaders.com/?q=privatebin.net&followRedirects=on and https://report-uri.com/home/analyse/https%3A%2F%2Fprivatebin.net%2F don't quite help here right now.

**Comment by rugk** [comment]
BTW on mobile so did not check that, but https://snip.dssr.ch/?72eaf59fb05b8584#7h27hKxFQngvML2ELCd1MuKc2Fg15Uv6fL1TrB4Zap42 fails for me too, so its maybe not actually related to the CSP. It has to be something else.

**Comment by elrido** [comment]
I tried it with some other browsers and the same browser on other devices - it doesn't work there either, but still works on the device I had originally tested the different headers I suspected. It's one of those bugs. :-/

So, if it's not the CSP, it might be one of these other ones I had toyed with disabling or changing:
- x-frame-options
- x-xss-protection

**Comment by rugk** [comment]
Well… do you have a website/instance that works now? Then we could try narrowing it down?

On Chromium I get the error code `ERR_BLOCKED_BY_RESPONSE` at least…

Aaaaand… ahaaa it's `Cross-Origin-Embedder-Policy`!

https://stackoverflow.com/questions/70752770/helmet-express-err-blocked-by-response-notsameorigin-200
https://stackoverflow.com/questions/69832906/how-to-prevent-err-blocked-by-response-notsameoriginafterdefaultedtosameoriginby

I just wonder about this, because no SO question tackles the case of pure _linking_ to other websites. They are about fetching stuff and so on, where obviously one can get errors.
Also [even if you use `require-corp`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy), it is actually same-origin, is not it?

Ahhh, but maybe it's [`Cross-Origin-Opener-Policy`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy) actually? There is also `Cross-Origin-Opener-Policy: same-origin-allow-popups`, which sounds like a header, we should try…

**Comment by rugk** [comment]
If that does not work, we could try `Cross-Origin Resource Policy: same-site`, which is _less safe_ than `Cross-Origin Resource Policy: same-origin`. But "opener" sounds like the thing which could trigger this bug, rather than "resource"… :eyes: 

I still miss a warning in some doc that/what of these can break links hmm…
Or maybe our case is special, because we link to our own origin in a new tab with `target="_blank"` (i.e. browsing context, i.e. new origin??)?

Ah and [there's the same header with "embedder"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy), which agaaain, I have no idea what the difference is. (But I hope it's just about iframes etc.)

BTW for future reference, that header was introduced in https://github.com/PrivateBin/docker-nginx-fpm-alpine/pull/32.

**Comment by elrido** [comment]
Thank you for digging further. I could not get it to work with the Embedder one, but the `Cross-Origin-Opener-Policy` did have an effect. Switching to the allow-popups didn't work, only removing it or leave it in the default `unsafe-none` did let me open the link. To avoid the debacle of last time, where it seemed to work in my browser, but afterwards didn't because I had something cached, I tested it in several browsers (incl. mobile) via our sponsored browserstack account. You can validate it on https://snip.dssr.ch/?72eaf59fb05b8584#7h27hKxFQngvML2ELCd1MuKc2Fg15Uv6fL1TrB4Zap42

I'll raise an MR to remove that header (with a comment, for future reference) - leaving it set with the unsafe value would just increase the sent header size, while not adding anything security-wise.
